google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,474 @@
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (stream_state_ == State::kShutdown) {
+        return make_ready_future(kPermanentErrorStatus);
+      }
+      stream_state_ = State::kShutdown;
+    }
+
+    // Since `Finish` is being called, there must be no outstanding `Read` and
+    // `Write`. Along with the status not being `State::kShutdown`, this means
+    // that the internal stream is still valid, so we can call `Finish()` on it.
+    return stream_->Finish().then([this](future<Status> finish_future) {
+      auto finish_response = finish_future.get();
+      status_promise_.set_value(finish_response);
+      return make_ready_future(finish_response);
+    });
+  }
+
+ private:
+  enum class State { kRetrying, kInitialized, kShutdown };
+
+  void OnFailure(Status status) {
+    AttemptRetry(status, retry_policy_prototype_->clone(),
+                 backoff_policy_prototype_->clone());
+  }
+
+  void FinishOnStreamFail(
+      std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+          stream) {
+    stream->Finish().then([this](future<Status> finish_status) {
+      OnFailure(finish_status.get());
+    });
+  }
+
+  void SetReadWriteFutures() {
+    if (read_reinit_done_.has_value()) {
+      // want to set member variable to nullopt before setting future returned
+      // to caller
+      auto read_promise = read_reinit_done_;
+      read_reinit_done_ = absl::optional();
+      read_promise->set_value();
+    }
+    if (write_reinit_done_.has_value()) {
+      auto write_promise = write_reinit_done_;
+      write_reinit_done_ = absl::optional();
+      write_promise->set_value();
+    }
+  }
+
+  void AttemptRetry(Status status, std::shared_ptr<RetryPolicy> retry_policy,
+                    std::shared_ptr<BackoffPolicy> backoff_policy) {
+    bool shutdown;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      shutdown = stream_state_ == State::kShutdown;
+    }
+    if (!shutdown && !retry_policy->IsExhausted() &&
+        retry_policy->OnFailure(status)) {
+      sleeper_(backoff_policy->OnCompletion())
+          .then([this, retry_policy, backoff_policy](future<void> sleep_fu) {
+            Initialize();
+          });
+      return;
+    }
+
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      stream_state_ = State::kShutdown;
+    }
+    status_promise_.set_value(kPermanentErrorStatus);
+    SetReadWriteFutures();
+  }
+
+  void Initialize(std::shared_ptr<RetryPolicy> retry_policy,
+                  std::shared_ptr<BackoffPolicy> backoff_policy) {
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+        potential_stream;
+    future<bool> start_future;
+
+    potential_stream = stream_factory_();
+    start_future = potential_stream->Start();
+
+    auto start_future_result = start_future.then(
+        [this, potential_stream =
+                   std::move(potential_stream)](future<bool> start_future) {
+          if (!start_future.get()) {
```

Comment:
I think you're overcomplicating this. In the case when the start future fails, you know there are no outstanding read or write requests, you're already retrying.

Instead:

```
impl_->Start().then([this](future<bool> start_future) {
  if (!start_future.get()) {
    return impl_.Finish().then([this](future<Status> s){
        // There was a failure: Finish() should never be OK.
        return StatusOr<std::unique_ptr<
                    AsyncStreamingReadWriteRpc<RequestType, ResponseType>>>(s.get());
    });
  }
  return initializer_(std::move(impl_));
});
```